test/jdk/javax/swing/ImageIcon/MultiResolutionImageIcon.java

```diff
@@ -0,0 +1,127 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import javax.imageio.ImageIO;
+import javax.swing.ImageIcon;
+import java.awt.Color;
+import java.awt.Graphics;
+import java.awt.Image;
+import java.awt.image.BufferedImage;
+import java.awt.image.ImageObserver;
+import java.awt.image.ImageProducer;
+import java.awt.image.MultiResolutionImage;
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * @test
+ * @bug 8282526
+ * @summary Default icon is not painted properly
+ * @run main MultiResolutionImageIcon
+ */
+
+public class MultiResolutionImageIcon extends Image implements MultiResolutionImage {
+    BufferedImage img;
+
+    public static void main(String[] args) {
+        MultiResolutionImageIcon me = new MultiResolutionImageIcon();
+        me.test();
+    }
+
+    public MultiResolutionImageIcon() {
+        final String PATH_TO_FILE=MultiResolutionImageIcon.class.getResource("folder.png").getPath();
+        try {
+            img = ImageIO.read(new File(PATH_TO_FILE));
+        } catch (IOException ioe) {
+            throw new RuntimeException("Can't read image file. " + ioe);
+        }
+    }
+
+    public void test() {
+        ImageIcon icon = new ImageIcon();
+        icon.setImage(this);
+
+        BufferedImage test = new BufferedImage(16, 16, BufferedImage.TYPE_INT_ARGB);
+        Graphics g = test.createGraphics();
+        icon.paintIcon(null, g, 0, 0);
+        g.dispose();
+
+        int transfers = 0;
+        boolean last = true;
+        for (int i = 0; i < 16; i++) {
+            Color pixel = new Color(test.getRGB(12, i), true);
+            if (isWhite(pixel) ^ last) {
+                transfers++;
+                last = !last;
+            }
+        }
+        if (transfers < 8) {
+            try {
+                ImageIO.write(test, "png", new File("generated.png"));
+            } catch (IOException ignore) {}
+            throw new RuntimeException("Significant detail is lost in transition");
+        }
+    }
+
+    public boolean isWhite(Color c) {
+        return (c.getRed() + c.getGreen() + c.getBlue()) / 3 > 245;
+    }
+
+    @Override
+    public Image getResolutionVariant(double destImageWidth, double destImageHeight) {
+        return img;
+    }
+
+    @Override
+    public List<Image> getResolutionVariants() {
+        return Arrays.asList(img);
+    }
+
+
+    @Override
+    public int getWidth(ImageObserver observer) {
+        return 16;
+    }
+
+    @Override
+    public int getHeight(ImageObserver observer) {
+        return 16;
+    }
+
+    @Override
+    public ImageProducer getSource() {
+        return null;
+    }
+
+    @Override
+    public Graphics getGraphics() {
+        return img.getGraphics();
+    }
+
+    @Override
+    public Object getProperty(String name, ImageObserver observer) {
+        return null;
+    }
+}
```

Comment:
This test reads a little strange to me. (I'm not an official openjdk reviewer, though, so feel free to disregard.)

If I run this test with the existing ImageIcon class it creates an image that resembles:
![image](https://user-images.githubusercontent.com/7669569/158093894-be7a73c6-6936-4725-bb09-cd054a2dccc2.png)

If I run this test with the revised code it creates an image that resembles:
![image](https://user-images.githubusercontent.com/7669569/158093939-bffc62c7-7a5c-4b0e-b71e-710c5049daf6.png)

Based on the changes in ImageIcon: I only expected to see changes related to rendering hints. (That is: I expected to see differences in antialiasing and scaling artifacts.)

Would you mind if I proposed a different test?

Reply:
Ah, I figured out part of my confusion:

In my case (my primary monitor was set to 100% resolution): the ImageIcon we get back uses a BufferedImage that is 16x16. So when it is displayed on a 150% monitor it has to scale *up*.

Depending on your monitor resolution: you might start at a 32x32 BufferedImage that is scaling down. Or you might not have a BufferedImage at all: you might have their custom MultiResolutionIconImage (which is backed by a BufferedImage).

So all that is to say: I misunderstood the original code's design. And now it sounds like there a lot of separate convos here, so I'll make this my last comment unless someone wants to follow-up with anything.

Additional context/feedback:

The image in question is coming from Win32ShellFolder2#makeIcon(..)

```
<init>:315, BufferedImage (java.awt.image)
makeIcon:1020, Win32ShellFolder2 (sun.awt.shell)
call:1066, Win32ShellFolder2$15 (sun.awt.shell)
call:1039, Win32ShellFolder2$15 (sun.awt.shell)
run$$$capture:264, FutureTask (java.util.concurrent)
run:-1, FutureTask (java.util.concurrent)
 - Async stack trace
<init>:132, FutureTask (java.util.concurrent)
newTaskFor:108, AbstractExecutorService (java.util.concurrent)
submit:139, AbstractExecutorService (java.util.concurrent)
invoke:621, Win32ShellFolderManager2$ComInvoker (sun.awt.shell)
invoke:519, ShellFolder (sun.awt.shell)
invoke:505, ShellFolder (sun.awt.shell)
getIcon:1039, Win32ShellFolder2 (sun.awt.shell)
getSystemIcon:249, FileSystemView (javax.swing.filechooser)
main:19, SwingDemo
```

If we wanted to apply a resolution somewhere other than the ImageIcon class:

We could instead change makeIcon() to return a custom MultiResolutionImage that always applies interpolation. Here's a variation of the original bug's app that models this idea:

https://docs.google.com/document/d/1fcXJ9g6uTTZZlIt1BfzNFg2-Hs05CCpNRjyG1FnXKns/edit?usp=sharing

I could spin this up as a PR or discuss this further if requested. I'm not saying it's necessarily better; I'm just brainstorming other approaches.